File: packages/types/src/issue-type.ts
// Issue Type types

export type TIssueTypeLogoProps = {
  in_use: string;
  icon?: {
    name: string;
    color: string;
  };
  emoji?: {
    value: string;
    url: string;
  };
};

export interface IIssueType {
  readonly id: string;
  workspace_id: string;
  name: string;
  description: string;
  logo_props: TIssueTypeLogoProps;
  is_epic: boolean;
  is_default: boolean;
  is_active: boolean;
  level: number;
}

export interface IIssueTypeLite {
  readonly id: string;
  name: string;
  logo_props: TIssueTypeLogoProps;
  is_epic: boolean;
}

export interface IProjectIssueType {
  readonly id: string;
  project_id: string;
  issue_type_id: string;
  issue_type_detail?: IIssueTypeLite;
  level: number;
  is_default: boolean;
}

export type TIssueTypeOperationsCallbacks = {
  createIssueType: (data: Partial<IIssueType>) => Promise<IIssueType>;
  updateIssueType: (issueTypeId: string, data: Partial<IIssueType>) => Promise<IIssueType | undefined>;
  deleteIssueType: (issueTypeId: string) => Promise<void>;
};

export type TProjectIssueTypeOperationsCallbacks = {
  addIssueType: (issueTypeId: string) => Promise<IProjectIssueType>;
  updateProjectIssueType: (
    projectIssueTypeId: string,
    data: Partial<IProjectIssueType>
  ) => Promise<IProjectIssueType | undefined>;
  removeIssueType: (projectIssueTypeId: string) => Promise<void>;
  setDefaultIssueType: (projectIssueTypeId: string) => Promise<void>;
};

File: apps/api/plane/app/serializers/issue_type.py
# Module imports
from .base import BaseSerializer
from rest_framework import serializers

from plane.db.models import IssueType, ProjectIssueType


class IssueTypeSerializer(BaseSerializer):
    class Meta:
        model = IssueType
        fields = [
            "id",
            "workspace_id",
            "name",
            "description",
            "logo_props",
            "is_epic",
            "is_default",
            "is_active",
            "level",
        ]
        read_only_fields = ["workspace_id"]


class IssueTypeLiteSerializer(BaseSerializer):
    class Meta:
        model = IssueType
        fields = [
            "id",
            "name",
            "logo_props",
            "is_epic",
        ]
        read_only_fields = fields


class ProjectIssueTypeSerializer(BaseSerializer):
    issue_type_detail = IssueTypeLiteSerializer(source="issue_type", read_only=True)

    class Meta:
        model = ProjectIssueType
        fields = [
            "id",
            "project_id",
            "issue_type_id",
            "issue_type_detail",
            "level",
            "is_default",
        ]
        read_only_fields = ["project_id"]
